File: archives/C#/0134. Gas Station.cs
public class Solution {
    public int CanCompleteCircuit(int[] gas, int[] cost) {
        for(int i=0;i<gas.Length;i++){
            if(CanCompleteCircuit(gas,cost,i))
                return i;
        }
        return -1;
    }
    public bool CanCompleteCircuit(int[] gas, int[] cost,int k){
        if(gas[k]<cost[k])
            return false;
        int len=gas.Length-1;
        int avgGas=0;
        while(k<gas.Length){
            avgGas=avgGas+gas[k]-cost[k];
            if(avgGas>=0){
                k++;
                len--;
            }
            else
                return false;
        }
        
        k=0;
        while(len>=0){
            avgGas=avgGas+gas[k]-cost[k];
            if(avgGas>=0){
                k++;
                len--;
            }
            else
                return false;
        }
        return true;
    } 
}

File: archives/leetcode2/0134. Gas Station.py
class Solution:
    def canCompleteCircuit(self, gas, cost):
        """
        :type gas: List[int]
        :type cost: List[int]
        :rtype: int
        """
        if sum(cost)>sum(gas):
            return -1
        for k in range(len(gas)):
            res=self.subcanCompleteCircuit(gas+gas,cost+cost,k,len(gas)+k)
            if res[0]:
                return res[1]
        return -1
    
    def subcanCompleteCircuit(self,gas,cost,start,end):
        rep=0
        for k in range(start,end+1):
            rep+=gas[k]-cost[k]
            if rep<0:
                return False,-1
        return True,start
            
